### DISCUSSION TITLE
How do I bring helpers into scope?

### DISCUSSION THREAD

**Discussion by Kodlak15** [nix-community/nixvim#1283] [Q&A] [answered]
This feels like a bit of a silly question, but I am clearly missing something here. I am using the home-manager Nixvim module and would like to pass some raw Lua code using a keybinding in Telescope. It seems as though the helpers.mkRaw function can be used to achieve this, so I followed the instructions in the documentation and tried loading helpers in my telescope.nix file like so:
``
{helpers, ...}: {
  ...
}
``
When I run my rebuild script I get an error message that mentions a missing 'helpers' attribute. Is there an additional step not mentioned in the documentation that is required to bring 'helpers' into scope?

**Comment by careb0t** [comment]
Did you ever figure this out? It seems based on your description that the docs might have slightly changed, but I am also getting a missing 'nixvim' attribute error trying to do the same exact thing. Just need to use some raw lua code for one of my plugins.

**Comment by MattSturgeon** [comment, marked as answer]
There's two ways to approach this.


## Using a submodule definition

You can assign a module-function to `programs.nixvim` and that module function has access to nixvim's modules args.

```nix
{
  programs.nixvim =
    {
      lib,
      ...
    }:
    {
      enable = true;
      someOption = lib.nixvim.mkRaw ''
        print('Hi!');
      '';
    };
}
```

## Via the `lib` option

Alternatively, we make most of our lib extensions available to the "host" modules (in this case, home-manager) via the `lib` config option.

Note that the `lib` config option is entirely unrelated to the `lib` module arg.

```nix
{ lib, config, ... }:
let
  # Access our helpers via config.lib
  helpers = config.lib.nixvim;
in
{
  programs.nixvim.someOption = helpers.mkRaw ''
    print('Hi!');
  '';
}
```

As @careb0t mentioned, the docs were recently updated to better reflect this: https://nix-community.github.io/nixvim/user-guide/helpers.html

**Reply by careb0t** [reply]
Thanks for the reply @MattSturgeon ! I'm so glad your response was so quick haha. I've imported nixvim from a flake into my home-manager `home.nix` file as seen in the screenshots below based on your response, but I am still receiving a `missing 'nixvim' attribute` error from the `helpers = config.lib.nixvim` line at the top of my `home.nix` file. According to my understanding of the documentation, this is the required method for using the helpers if nixvim is accessed via flake. Can you tell what I am doing wrong here?

_home.nix_
![image](https://github.com/user-attachments/assets/d2090c95-04ff-4d84-8e3b-e3b6d76cda2f)
![image](https://github.com/user-attachments/assets/b344ae0a-9b8f-4a83-a2c7-4ccec849b815)

_flake.nix_
![image](https://github.com/user-attachments/assets/40c2e611-ce5b-4fed-9511-20a75d71d78e)

**Reply by MattSturgeon** [reply]
You're using nixvim 24.05, which is lagging behind our "main" branch. That's why this isn't working for you.

The "stable" version of our docs are [here](https://nix-community.github.io/nixvim/stable/user-guide/helpers.html).

You should be able to use something like `config.nixvim.helpers` until you upgrade to 24.11.

**Reply by MattSturgeon** [reply]
As an aside, my first method of assigning a function-module to `programs.nixvim` will still work, but back in our 24.05 version we didn't yet have `lib.nixvim` so you'd have to use `helpers` instead.
